server/src/main/java/de/uftos/repositories/solver/timefold/constraints/ConstraintDefinitionFactory.java

```diff
@@ -0,0 +1,1073 @@
+        convertElement(equation.parameters().getLast(), params);
+    return (parameters) -> {
+      ResourceTimefoldInstance first = firstElement.apply(parameters);
+      ResourceTimefoldInstance second = secondElement.apply(parameters);
+
+      if (first.getResourceType() != ResourceType.NUMBER
+          || second.getResourceType() != ResourceType.NUMBER) {
+        throw new IllegalArgumentException();
+      }
+
+      return ((Number) first).getValue() > ((Number) second).getValue();
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, Boolean> convertIn(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() != UcdlToken.IN) {
+      throw new IllegalStateException();
+    }
+    OperatorDto in = (OperatorDto) ast;
+    if (in.parameters().size() != 2) {
+      throw new IllegalStateException();
+    }
+    Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> element =
+        convertElement(in.parameters().getFirst(), params);
+    Function<List<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>> set =
+        convertSet(in.parameters().getLast(), params);
+    return (parameters) -> set.apply(parameters).contains(element.apply(parameters));
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>> convertOf(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() != UcdlToken.OF) {
+      throw new IllegalStateException();
+    }
+
+    OperatorDto of = (OperatorDto) ast;
+
+    if (of.parameters().size() != 2) {
+      throw new IllegalStateException();
+    }
+
+    AbstractSyntaxTreeDto set = of.parameters().getLast();
+
+    return convertSet(set, params);
+  }
+
+  private static Function<
+      List<ResourceTimefoldInstance>,
+      Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+      > convertFilter(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() != UcdlToken.FILTER) {
+      throw new IllegalStateException();
+    }
+
+    OperatorDto filter = (OperatorDto) ast;
+
+    List<Function<List<ResourceTimefoldInstance>, Boolean>> functions = new ArrayList<>();
+
+    for (AbstractSyntaxTreeDto param : filter.parameters()) {
+      if (param.getToken() == UcdlToken.NUMBER_SET || param.getToken() == UcdlToken.RESOURCE_SET) {
+        Function<List<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>> set =
+            convertSet(param, params);
+        Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> element =
+            convertValueReference(new ValueDto<>(UcdlToken.VALUE_REFERENCE, "this"), params);
+        functions.add((parameters -> set.apply(parameters).contains(element.apply(parameters))));
+      } else {
+        functions.add(convertBool(param, params));
+      }
+    }
+
+    return (parameters -> set -> {
```

Comment:
What is happening there? That's a function taking parameters and mapping to another function that takes `set` as an argument?

Reply:
Basically it is a method which returns a function which returns a function.
It could also be seen as a method which returns a function, that takes 2 parameters.